api/include/opentelemetry/logs/logger.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+#include <chrono>
+#include <map>
+#include <vector>
+
+#include "opentelemetry/common/key_value_iterable.h"  // fix after moving to "common"
+#include "opentelemetry/logs/log_record.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace logs
+{
+/**
+ * Handles log record creation.
+ *
+//  */
+// static const struct LogRecord defaultRecord = {
+//   timestamp : std::chrono::steady_clock::now().time_since_epoch(),
+//   trace_id : 0,
+//   span_id : 0,
+//   severity_number : 0
+// };
+
+class Logger
+{
+private:
+  std::vector<LogRecord> records;
+
+public:
+  virtual ~Logger() = default;
+
+  /* Returns the name of the logger */
+  virtual nostd::string_view getName() = 0;
+
+  /**
+   * Creates a log message with the passed characteristics.
+   *
+   * @param name the name of the log event.
+   * @param sev the severity level of the log event.
+   * @param msg the string message of the log (perhaps support std::fmt or fmt-lib format).
+   * @param record the log record (object type LogRecord) that is logged.
+   * @param attributes the attributes, stored as a 2D list of key/value pairs, that are associated
+   * with this log.
+   * @param some_obj the log object of user templated type that is logged.
+   * @throws No exceptions under any circumstances.
+   */
+
+  /* A logging statement that takes in a LogRecord.
+   * A default LogRecord that will be assigned if no parameters are passed to Logger's .log() method
+   * which should at minimum contain the trace_id, span_id, and current timestamp
+   * TODO: at the minimum, correlate timestamp, traceid, and spanid to Log Record by default
+   *
+   */
+  virtual void log(const LogRecord &record) noexcept = 0;
+
+  /** Overloaded methods for an unstructured logging **/
+  void log(nostd::string_view msg) noexcept
+  {
+    log(Severity::NONE, msg);  // Set severity to NONE as default then call the log method below
+  }
+
+  void log(Severity sev, nostd::string_view msg) noexcept
```

Comment:
@xukaren @MarkSeufert - thoughts (inspired by looking at C# ILogger and log4j).

We may provide header only inline methods in logger.h:
(pseudocode)

```cpp
    void debug(...) -> log(DEBUG, ...)
    void error(...) -> log(ERROR, ...)
    void fatal(...) -> log(FATAL, ...)
    void info(...) -> log(INFO, ...)
    void warn(...) -> log(WARN, ...)
    void trace(...) -> log(TRACE, ...)
```
Assuming here that OpenTelemetry spec order of these matches the conventional order used by log4j and log4cxx.

Practical considerations:

Most apps and services that are already instrumented with whatever logging framework today would have the same primitives (log4j, log4cxx, C# ILogger), and/or similarly structured methods with just that set. While we still retain flexibility here -- `log(severity, ...)` provides fine-grained logging control with 20 levels or alike, we still need to maintain _convenience_  / similarity to existing frameworks. From performance standpoint - this is compiling into same code, but `shortcuts` above allow for more natural, idiomatic experience, that devs have already seen / used to in other logging frameworks.

I think adding those overloads with substituting param would be nice, even if we only express the core set with these helper / convenience / "syntactic sugar" methods.

Reply:
Another variant of naming convention in C# ILogger - is:
```
    LogCritical(...)
    LogDebug(...)
    LogError(...)
    LogInformation(...)
    LogTrace(...)
    LogWarning(...)
```

Please check what methods names we have in the other parts... Maybe camelCased function 'aliases' for convenience? e.g. `logDebug, logError, logFatal, logInfo, logWarn, logTrace`

Reply:
Before being able to support the set of convenience functions (`void debug(..)`, `void trace(..)`, etc.), wouldn;t we first have to decide the most likely order and combination of parameters so we are not changing the `debug(..)`/`trace(..)`/etc. methods as we go along changing the API. But eventually I think they should be added. 

It should also be easy to add function aliases once we figure out the convention as well.

Also should we only support the 'convenience functions' for the main severity levels? (i.e. only `void debug(..)` not `void debug2(..)`, `void debug2(..)` etc.)

On another note, do you have any opinions on whether methods such as "`getCallingMethodName()` or `getLineNumber()` would be useful for users?

Reply:
We should discuss the order. Yes.

I am not sure we need to have the methods, but it may be helpful to have a logging macro that may capture current scope it context, functionally similar to log4j MDC, that carries info such as function/method, module or LOG_TAG, file, line number. All of these may go to whatever attribs in context, in a property bag we pass to SDK. This is separate topic we need to discuss after we finish the basic structured logging.